src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -581,16 +578,16 @@ public sealed override void WritePrimitive(ODataPrimitiveValue primitiveValue)
             this.WritePrimitiveValueImplementation(primitiveValue);
         }
 
-
         /// <summary>
         /// Asynchronously write a primitive value.
         /// </summary>
         /// <param name="primitiveValue"> Primitive value to write.</param>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
-        public sealed override Task WritePrimitiveAsync(ODataPrimitiveValue primitiveValue)
+        public sealed override async Task WritePrimitiveAsync(ODataPrimitiveValue primitiveValue)

```

Comment:
Why are we not eliding async in this method, like the way we are doing in `WriteDeltaDeletedLinkAsync` ?

Same question for the methods below:
`WriteStartAsync`
`CreateBinaryWriteStreamAsync`
`CreateTextWriterAsync`
`WriteEntityReferenceLinkAsync`
`WriteEntityReferenceLinkAsync`

Reply:
Also
I can see we have some `Verify****Async` methods. Why are some missing?

Reply:
@KenitoInc The reason we're not eliding is because the statement `this.VerifyCanWritePrimitive(false, primitiveValue);` can throw an exception. By making the method async, a state machine will be generated and such an exception will be captured and placed on the returned task. If we elide `async`, the exception is raised directly rather than going on the task. The expected asynchronous semantics are that exceptions are placed on the returned task.
For that same reason I have applied `async` to `WriteDeltaDeletedLinkAsync`.

Regarding your second question about missing `Verify***Async` methods, the only scenarios where I introduced an asynchronous equivalent is where the verify method does an asynchronous operation. For example,
```csharp
        private async Task VerifyCanWriteStartResourceSetAsync(bool synchronousCall, ODataResourceSet resourceSet)
        {
            ExceptionUtils.CheckArgumentNotNull(resourceSet, "resourceSet");

            this.VerifyNotDisposed();
            this.VerifyCallAllowed(synchronousCall);

            await this.StartPayloadInStartStateAsync()
                .ConfigureAwait(false);
        }
```
As you can see here, this method is initiating an asynchronous operation to start writing the payload. One apparent observation would be that the verify method is doing a little more than it'd be expected to but the primary objective of this initiative is to achieve parity with the synchronous API.
The alternative approach would be to move the call to `StartPayloadInStartState` from `VerifyCanWriteStartResourceSet` to the calling method such that the verify method can be reused by both the synchronous and asynchronous workflow but I thought I could do that in a refactor PR.